aidial_assistant/utils/open_api.py

```diff
@@ -0,0 +1,75 @@
+from typing import Any, Iterable, Tuple
+
+from langchain_community.utilities.openapi import OpenAPISpec
+from openai.types.chat import ChatCompletionToolParam
+from openapi_pydantic import DataType, Reference, Schema
+
+from aidial_assistant.utils.open_ai import construct_tool
+
+
+def _resolve_schema(spec: OpenAPISpec, schema: Schema | Reference) -> Schema:
+    if isinstance(schema, Reference):
+        return spec.get_referenced_schema(schema)
+
+    return schema
+
+
+def _construct_property(
+    spec: OpenAPISpec, schema: Schema | Reference
+) -> dict[str, Any]:
+    return _resolve_schema(spec, schema).dict(exclude_none=True)
+
+
+def _extract_body_parameters(
+    spec: OpenAPISpec, schema: Schema | Reference
+) -> Iterable[Tuple[str, dict[str, Any], bool]]:
+    schema = _resolve_schema(spec, schema)
+    if schema.type != DataType.OBJECT:
+        raise ValueError("Body schema must be an object")
+
+    if schema.properties:
+        required = schema.required or []
+        for prop_name, prop_schema in schema.properties.items():
+            prop_schema = _resolve_schema(spec, prop_schema)
+
+            yield (
+                prop_name,
+                _construct_property(spec, prop_schema),
+                prop_name in required,
+            )
+
+
+def construct_tool_from_spec(
+    spec: OpenAPISpec, path: str, method: str
+) -> ChatCompletionToolParam:
+    operation = spec.get_operation(path, method)
+    properties: dict[str, Any] = {}
+    required = []
+    for p in spec.get_parameters_for_operation(operation):
+        if p.param_schema is None:
+            raise ValueError(f"Parameter {p.name} has no schema")
+
+        properties[p.name] = _construct_property(spec, p.param_schema)
+
+        if p.required:
+            required.append(p.name)
+
+    request_body = spec.get_request_body_for_operation(operation)
+    if request_body is not None:
+        for key, media_type in request_body.content.items():
```

Comment:
You could lookup the dictionary `media_type = request_body.content.get("application/json")` instead of looping through all items.